modules/pool/lib/staking/gauge-staking.service.ts

```diff
@@ -1,221 +1,252 @@
+                this.balMulticaller.call(`${gauge.id}.workingSupply`, gauge.id, 'working_supply', [], true);
             }
-        });
+        }
+        const balData = await this.balMulticaller.execute() as GaugeBalDistributionData;
 
-        const childChainData = (await multicall.execute()) as Record<string, string | undefined>;
-
-        for (const childChainGauge in childChainData) {
-            if (childChainData[childChainGauge]) {
-                response[childChainGauge] = {
-                    version: 2,
-                    rate: formatUnits(childChainData[childChainGauge]!, 18),
-                };
-            } else {
-                response[childChainGauge] = {
-                    version: 1,
-                    rate: '0.0',
-                };
+        // Get onchain data for reward tokens
+        for (const gauge of gauges) {
+            for (const token of gauge.tokens ?? []) {
+                const [address] = token.id.toLowerCase().split('-');
+                if (gauge.version === 1) {
+                    this.rewardsMulticallerV1.call(`${gauge.id}.rewardData.${address}`, gauge.id, 'reward_data', [address], true);
+                } else {
+                    this.rewardsMulticallerV2.call(`${gauge.id}.rewardData.${address}`, gauge.id, 'reward_data', [address], true);
+                }
             }
         }
+        const rewardsDataV1 = await this.rewardsMulticallerV1.execute() as GaugeRewardData;
+        const rewardsDataV2 = await this.rewardsMulticallerV2.execute() as GaugeRewardData;
+        const rewardsData = { ...rewardsDataV1, ...rewardsDataV2 };
+
+        const totalBalRate = emissions.weekly() / 604800;
+        const now = Math.floor(Date.now() / 1000);
+
+        // Format onchain rates for all the rewards
+        const onchainRates = [
+            ...Object.keys(balData).map((gaugeAddress) => ({
+                id: `${this.balAddress}-${gaugeAddress}`.toLowerCase(),
+                rewardPerSecond: balData[gaugeAddress]?.rate
+                    ? formatUnits(balData[gaugeAddress].rate!) // L2 V2 case
+                    : (parseFloat(formatUnits(balData[gaugeAddress].weight!)) * totalBalRate).toFixed(18), // mainnet case
+                workingSupply: balData[gaugeAddress]?.workingSupply ? formatUnits(balData[gaugeAddress].workingSupply!) : '0',
+            })),
+            ...Object.keys(rewardsData).map((gaugeAddress) => [ // L2 V1 case, includes tokens other than BAL
+                ...Object.keys(rewardsData[gaugeAddress].rewardData).map((address) => ({
+                    id: `${address}-${gaugeAddress}`.toLowerCase(),
+                    rewardPerSecond: rewardsData[gaugeAddress].rewardData[address]?.period_finish
+                        && rewardsData[gaugeAddress].rewardData[address]?.period_finish!.toNumber() > now
+                        && formatUnits(rewardsData[gaugeAddress].rewardData[address].rate!)
+                        || '0.0',
```

Comment:
Why would this be a string and not a boolean? What do I miss?

Reply:
not sure I understand, do you mean the logic? It translates to: if period_finish was fetched and it's in the future then use the formatted rate otherwise set it to 0.

Reply:
zeros are filtered out later on, so we don't have store empty values in the DB